File: js/이동하기(11408).js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);

const miro = [];
for (let i=1; i<=n; i++) {
    miro.push(input[i].split(' ').map(Number));
}

// 전치 행렬
const miro_reverse = Array.from({length:m}, () => Array(n).fill(0));
for (let i=0; i<n; i++) {
    for (let j=0; j<m; j++) {
        miro_reverse[j][i] = miro[i][j];
    }
}

const dp = Array.from({ length: n }, () => Array(m).fill(0));
dp[0][0] = miro[0][0];

for (let i = 1; i < m; i++) {
    dp[0][i] = dp[0][i-1] + miro[0][i];
}

for (let j = 1; j < n; j++) {
    dp[j][0] = dp[j-1][0] + miro_reverse[0][j];
}

for (let i = 1; i < n; i++) {
    for (let j = 1; j < m; j++) {
        dp[i][j] = Math.max(dp[i-1][j], dp[i][j-1], dp[i-1][j-1]) + miro[i][j];
    }
}

console.log(dp[n-1][m-1]);

File: python/이동하기(11408).py
n, m = map(int, input().split())
miro = [
    list(map(int, input().split()))
    for _ in range(n)
]

dp = [[0] * m for _ in range(n)]

miro_reverse = list(row for row in zip(*miro))

dp[0][0] = miro[0][0]

for i in range(1, m):
    dp[0][i] = sum(miro[0][:i+1])

for j in range(1, n):
    dp[j][0] = sum(miro_reverse[0][:j+1])

for i in range(1, n):
    for j in range(1, m):
        dp[i][j] = max(dp[i-1][j], dp[i][j-1], dp[i-1][j-1]) + miro[i][j]

print(dp[n-1][m-1])

